fs: Implement fs_mkfs operation

fkokosinski opened the pull request:
This PR implements a new API call for the file subsystem: `fs_mkfs`. It triggers the format operation in the underlying file system.

This PR also adds a new sample that uses this call in two distinct scenarios.

de-nordic reviewed (changes requested):
Separate changes to commits by subject:
1) API change
2) FATFS change
3) LittleFS change
4) sample

The change does not respect `CONFIG_FS_FATFS_MOUNT_MKFS` and `CONFIG_FS_FATFS_READ_ONLY `.

`fs_mkfs` should not be required API call, so should be optional and configured by Kconfig.

fzdobylak commented:
@de-nordic Thank you for the review!

I will change my code to respect FATFS flags.

> `fs_mkfs` should not be required API call, so should be optional and configured by Kconfig.

Why should this function be optional? I think that this is a common operation and there is no need to load it conditionally. I could make it optional in FATFS where `mkfs` operation is truly optional, but in littlefs that operation is always available (but it is triggered only if the file system is not detected).

de-nordic commented:
> 
> Why should this function be optional? I think that this is a common operation and there is no need to load it conditionally. 

How often do you format your storage?

I know use cases where the FAT FS or LittleFS partitions are generated at compile time and merged into hex, as they contain GUI or other resources used by applications; this is not a perfect use but there is no write operations used there, nor anyone wants to format anything when mount fails.

> I could make it optional in FATFS where `mkfs` operation is truly optional, but in littlefs that operation is always available (but it is triggered only if the file system is not detected).

And this is actually not good thing, because this means that if something fails you can get your data on LittleFS device destroyed for no reason, and is not entirely true, search for `FS_MOUNT_FLAG_NO_FORMAT`.

I am not a fan of formatting device when you can not access it, by default. I have already experienced data loss when I was unfortunate to plug SD card with MBR and two partitions into a device that did not recognize my SD and got it wiped, while I was just trying to move some data onto that device.

fkokosinski commented:
@de-nordic hi, could you take a look at this PR again?

fkokosinski commented:
Hi @nvlsianpu, as you are the assigned reviewer to this PR, could you perhaps take a look at it?

de-nordic commented:
> @de-nordic hi, could you take a look at this PR again?

Sorry, got crushed with other tasks. Will try to look at this today.

de-nordic commented on subsys/fs/fat_fs.c:
```diff
@@ -458,6 +458,31 @@ static int fatfs_unmount(struct fs_mount_t *mountp)
+
+static int fatfs_mkfs(struct fs_mount_t *mountp)
+{
+#if defined(CONFIG_FS_FATFS_MOUNT_MKFS)
```
This is not correct, the Kconfig used here is to enable/disable mkfs during mount, it should not affect mkfs, which is independent from mount.

fzdobylak replied:
I have removed that.

de-nordic commented on subsys/fs/fs.c:
```diff
@@ -720,6 +720,54 @@ int fs_mount(struct fs_mount_t *mp)
+		goto mount_err;
+	}
+
+	rc = fs->mkfs(mp);
```
mkfs should respect `FS_MOUNT_FLAG_NO_FORMAT`, if such flag is set to mount point  mkfs should return `-EROFS`.

fzdobylak replied:
Now I also check for that flag.

de-nordic commented:
Add release notes entry for 3.3, as additional commit, with information on added Kconfig option.

fzdobylak commented on subsys/fs/littlefs_fs.c:
```diff
@@ -860,6 +881,56 @@ static int littlefs_mount(struct fs_mount_t *mountp)
+	ret = littlefs_init_cfg(fs, block_dev);
+	if (ret < 0) {
+		goto out;
+	}
```
I have created a `littlefs_init_mp` function that gathers most of the common code.

fzdobylak commented on subsys/fs/fs.c:
```diff
@@ -720,6 +720,54 @@ int fs_mount(struct fs_mount_t *mp)
+		LOG_ERR("requested file system type not registered!!");
+		rc = -ENOENT;
+		goto mount_err;
+	}
```
I have created a function `validata_mp` that does all the checks found in `fs_mount`. You are right that all these checks should be also done in `fs_mkfs` so I use this new function also there.

fzdobylak commented:
@de-nordic Thank you for your review. Could you check now if everything is correct?

de-nordic commented on subsys/fs/fs.c:
```diff
@@ -721,6 +733,51 @@ int fs_mount(struct fs_mount_t *mp)
+
+	rc = fs->mkfs(mp);
+	if (rc < 0) {
+		LOG_ERR("fs mount error (%d)", rc);
```
not a mount error

fzdobylak replied:
fixed

de-nordic commented on samples/subsys/fs/format/README.md:
```diff
@@ -0,0 +1,33 @@
+# Format Sample Application
+
+## Overview
+This sample is designed to show how to format different storage devices for different
```
"The aim of this sample application is to..."

fzdobylak replied:
Updated.

de-nordic reviewed (changes requested):
Needed addition of `mkfs` tests here: zephyr/tests/subsys/fs/fs_api/

de-nordic commented on subsys/fs/littlefs_fs.c:
```diff
@@ -860,6 +897,42 @@ static int littlefs_mount(struct fs_mount_t *mountp)
+	struct fs_littlefs *fs = mountp->fs_data;
+
+	/* Create and take mutex. */
+	k_mutex_init(&fs->mutex);
+	fs_lock(fs);
```
What will happen here when LittleFS at given mount point is already mounted, which means that it has already been initialized?

de-nordic replied:
Ah sorry, this is covered by `validate_mp`

fzdobylak commented on samples/subsys/fs/format/prj.conf:
```diff
@@ -0,0 +1,12 @@
+CONFIG_DEBUG=y
```
Yes, it is unnecessary. Removed.

fzdobylak commented on samples/subsys/fs/format/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(format)
```
Updated.

fzdobylak commented on include/zephyr/fs/fs_sys.h:
```diff
@@ -38,6 +38,7 @@ extern "C" {
  * @param stat Checks the status of a file or directory specified by the path
  * @param statvfs Returns the total and available space on the file system
  *        volume
+ * @param mkfs Formats the device to contain fresh new file system
```
Updated.

de-nordic commented:
> 
> I think the better place to create new tests is a directory of tests for a real file system (e.g. LittleFS). Then we are able to make some stronger checks (e.g. if the file system was formatted or not). I think that the test for fs_mkfs will be identical for any file system hence I will add it to the common directory and use it in LittleFS and FAT.
> 
> If you have some suggestions or you disagree with my suggestion please clarify how the new tests should look like.

You have added new API call, so we need to have it tested.
It does not have to be just under fs_api, but what I think the test should do is (on native_posix target, RAM flash simulator):
1) zero the flash partition;
2) try to mount the partition with MKFS on mount disabled;
3) do the mkfs on the partition;
4) try to mount the partition again  - SUCCESS

Would it be possible to provide such simple scenario?

fzdobylak commented:
I have created that simple scenario and some other that checks if file system was formatted when there was previously some other file system on storage device. For now I have prepared that only for LittleFS. Could you tell if this is what you wanted?

de-nordic commented:
> I have created that simple scenario and some other that checks if file system was formatted when there was previously some other file system on storage device. For now I have prepared that only for LittleFS. Could you tell if this is what you wanted?

Yes. Although you can shorten it. You are testing mkfs so it should be enough to mkfs and then mount without error.

fzdobylak commented:
I have added tests to the FAT file system too.

de-nordic commented on subsys/fs/fs.c:
```diff
@@ -679,11 +673,29 @@ int fs_mount(struct fs_mount_t *mp)
 		}
 	}
 
+	if (length) {
```
`validate_mp` is internal function, so just require `length` to be not null with assert at the beginning, get rid of the check and make sure that `fs_mkfs` provides it with pointer, even if it discards the value.

fzdobylak replied:
In `fs_mkfs` we don't need the length of a mount point. I decided to make the length argument optional. I think that making this argument required is worse solution (then we will have to pass pointer to some unused variable).

de-nordic replied:
This argument is passed whether you give here a pointer or NULL, which is pointer. 
You have just added additional, redundant, check in internal only function, which will be check for no reason, other than make it possible for the never omitted argument to be null.

fzdobylak replied:
But why is creating an unused variable better? Here we have to make a choice of option that will be less confusing for someone who will read this in the future.

I think that passing a pointer to an unused variable will cause more confusion. The reader will spend some time wondering why we pass here a pointer to that variable. On the other hand if he sees a NULL pointer there, he will think “okay, we don’t care about the length of the mp”.

de-nordic commented on subsys/fs/fs.c:
```diff
@@ -721,6 +733,51 @@ int fs_mount(struct fs_mount_t *mp)
+		goto mount_err;
+	}
+
+	/* Get file system information */
+	fs = fs_type_get(mp->type);
+	if (fs == NULL) {
+		LOG_ERR("requested file system type not registered!!");
+		rc = -ENOENT;
+		goto mount_err;
+	}
```
This part can also be taken under `validate_mp`, as this is the same code as in `fs_mount`.
Can you also change the message to print the inf code for the type? I think it is missing that information.

fzdobylak replied:
I don't think that getting file system information should be a part of `validate_mp`. It is very different operation then checking if mount point is valid. Functions that do one thing are easier to understand hence I don't want to do anything other then checking mount point in that function. I suggest to leave `fs_type_get` where it is.

fzdobylak replied:
I have added information about a mp type to the error message.

de-nordic replied:
> I don't think that getting file system information should be a part of `validate_mp`. It is very different operation then checking if mount point is valid. Functions that do one thing are easier to understand hence I don't want to do anything other then checking mount point in that function. I suggest to leave `fs_type_get` where it is.

Is mount point object valid when  `mp->type` is ininvalid?

The  `validate_mp` only validates `mp` object in context of Zephyr subsystem, you are still not warranted to mount the file system, because, for example, FAT FS will only mount at specified mount points.
https://github.com/zephyrproject-rtos/fatfs/blob/master/include/ffconf.h#L204
This is probably thing for the next PR...

fzdobylak replied:
The purpose of the `validate_mp` function is to check if mp is valid in the sense of Zephyr subsystem. We cannot check if mp is valid for FAT FS if we don’t have retrieved file system information.

The file system type may be an arbitrary number (see: https://github.com/zephyrproject-rtos/zephyr/blob/07de2ffbb01a4633cdec7e48190ae369ecc37c9b/include/zephyr/fs/fs.h#L42:L47). Then it is impossible to check if that field is valid without trying to access it.

Maybe I will describe it in an analogy to the file system. One thing is to check if the file path is valid (consists of allowed characters). The other thing is looking for a file in the file system. If a file is not found then it doesn’t mean that path was invalid. It just means that file does not exist.

If you still think that `fs_type_get` should be inside `validate_mp` I will add another argument to that function that will be used to return the fs information.

de-nordic replied:
> The purpose of the `validate_mp` function is to check if mp is valid in the sense of Zephyr subsystem. We cannot check if mp is valid for FAT FS if we don’t have retrieved file system information.

You did not get my point. I do not expect you to validate mount point for the FAT FS or any other system. What you check here is validity of mp object in Zephyr context.

> 
> The file system type may be an arbitrary number (see: https://github.com/zephyrproject-rtos/zephyr/blob/07de2ffbb01a4633cdec7e48190ae369ecc37c9b/include/zephyr/fs/fs.h#L42:L47). Then it is impossible to check if that field is valid without trying to access it.

This argument makes no sense.  All systems are registered with `fs_register`, that includes Zephyr supported FAT and LittleFS. User may `fs_register` another FAT driver with new identifier (`fs_type`) and use it with `fs_mount_t` object definition.
VFS subsystem holds own list of know fs types and `fs_type_get` searches the list for that type.
If `validate_mp` checks mount point object, in the Zephyr context, it can also check whether the object carries valid, already registered type.

> 
> Maybe I will describe it in an analogy to the file system. One thing is to check if the file path is valid (consists of allowed characters). The other thing is looking for a file in the file system. If a file is not found then it doesn’t mean that path was invalid. It just means that file does not exist.

OK, so we have different understanding of `validate_mp`: I understand it as validation of `mp` object, as object defining mount point with valid path, not duplicating already mounted path, containing containing valid valid `fs_type`, which is valid candidate for being added to `fs_mnt_list`, if successfully mounted by the given file system type.

> 
> If you still think that `fs_type_get` should be inside `validate_mp` I will add another argument to that function that will be used to return the fs information.

Yes, that was what I have been asking for.

Now, after all the arguing for the change I actually think that we wen the wrong way from the beginning, and that signature of `fs_mkfs` should be:
`int fs_mkfs(int fs_type, uintptr dev_id)`

The `fs_mkfs` should only process fs_type and maybe check whether dev_id is not NULL, all other processing should be dumped on mkfs specific for file system.

fzdobylak replied:
Good point. I will try to make the new signature for that function and it may became simpler.

de-nordic replied:
> Good point. I will try to make the new signature for that function and it may became simpler.

Wait, there is one problem with this: we do not handle records of "used devices", which means that you could mkfs on device that is mouted, something that your existing solution would not allow!

fzdobylak committed:
fs: Add fs_mkfs operation to fs api

Adds fs_mkfs function to fs api. It will allow to perform mkfs operation
in file systems added to Zephyr.

Signed-off-by: Franciszek Zdobylak <fzdobylak@internships.antmicro.com>

fzdobylak committed:
fs: Implement mkfs operation in littleFS

Provide an implementation of fs_mkfs operation in littleFS.

Signed-off-by: Franciszek Zdobylak <fzdobylak@internships.antmicro.com>

fzdobylak committed:
fs: Implement mkfs operation in FAT fs

Provide an implementation of fs_mkfs operation in FAT fs.

Signed-off-by: Franciszek Zdobylak <fzdobylak@internships.antmicro.com>

fzdobylak committed:
samples: fs: Add sample using fs_mkfs function

Provide a demo which uses newly added fs_mkfs function.

Signed-off-by: Franciszek Zdobylak <fzdobylak@internships.antmicro.com>

fzdobylak committed:
doc: release-notes: add file systems notes

This commit adds file systems release notes for Zephyr v3.3.0.

Signed-off-by: Franciszek Zdobylak <fzdobylak@internships.antmicro.com>

fzdobylak committed:
tests: fs: Add test of fs_mkfs

Test is added to common directory and used in LittleFS tests.

Signed-off-by: Franciszek Zdobylak <fzdobylak@internships.antmicro.com>

fzdobylak commented:
@de-nordic Could you take a look at the new implementation of `fs_mkfs`? I hope that it is what you suggested in previous comments.

de-nordic commented on subsys/fs/fat_fs.c:
```diff
@@ -458,6 +458,27 @@ static int fatfs_unmount(struct fs_mount_t *mountp)
 
+#if defined(CONFIG_FILE_SYSTEM_MKFS)
+
+static int fatfs_mkfs(uintptr_t dev_id, int flags)
```
One more suggestion: remove flags and use void * pointer to allow passing backend dependent init object, and providing default behaviour when the pointer to init object is null.
I am using the fatfs_mkfs as an example here, but I mean the change should be general.
`inf flags` gives you nothing and does not matter because you are directly asking subsystem to format a device for you.
For example when `fatfs_mkfs(dev, mkfs_opt)` gets called with NULL for mkfs_opt (the init object), then it will initialize the device as the code below does. But if user does the extra mile and passes own MKFS_PARM type object, the mkfs will format the device as requested.

de-nordic replied:
Similar thing should be possible for the LittleFS

fzdobylak commented:
@de-nordic Hi, could you revisit this PR once again?

de-nordic commented on include/zephyr/fs/fs.h:
```diff
@@ -605,6 +605,23 @@ int fs_stat(const char *path, struct fs_dirent *entry);
+ *
+ * @param fs_type Type of file system to create
+ * @param dev_id Id of storage device
+ * @param cfg Backend dependent init object
```
Note that this is optional and may be NULL, in which case the default configuration will be used.

de-nordic commented on subsys/fs/fs.c:
```diff
@@ -721,6 +721,42 @@ int fs_mount(struct fs_mount_t *mp)
+
+	rc = fs->mkfs(dev_id, cfg, flags);
+	if (rc < 0) {
+		LOG_ERR("fs mkfs error (%d)", rc);
```
"fs" not needed, "mkfs .. " is enough.

fzdobylak replied:
Same as above.

de-nordic reviewed (changes requested):
Minor comments, otherwise looks OK.

fzdobylak commented on subsys/fs/fs.c:
```diff
@@ -721,6 +721,42 @@ int fs_mount(struct fs_mount_t *mp)
+	/* Get file system information */
+	fs = fs_type_get(fs_type);
+	if (fs == NULL) {
+		LOG_ERR("requested file system type (%d) not registered!!",
```
Actually I have done that to be consistent with log messages in fs_mount function.

fzdobylak replied:
Should I change it also in `fs_mkfs` function?
